pkg/flow/internal/controller/loader.go

```diff
@@ -564,80 +568,103 @@ func (l *Loader) OriginalGraph() *dag.Graph {
+		return
+	}
 	tracer := l.tracer.Tracer("")
-
-	l.mut.RLock()
-	defer l.mut.RUnlock()
+	spanCtx, span := tracer.Start(context.Background(), "SubmitDependantsForEvaluation", trace.WithSpanKind(trace.SpanKindInternal))
+	span.SetAttributes(attribute.Int("originators_count", len(updatedNodes)))
+	defer span.End()
 
 	l.cm.controllerEvaluation.Set(1)
 	defer l.cm.controllerEvaluation.Set(0)
-	start := time.Now()
-
-	spanCtx, span := tracer.Start(context.Background(), "GraphEvaluatePartial", trace.WithSpanKind(trace.SpanKindInternal))
-	span.SetAttributes(attribute.String("initiator", c.NodeID()))
-	defer span.End()
 
-	logger := log.With(l.log, "trace_id", span.SpanContext().TraceID())
-	level.Info(logger).Log("msg", "starting partial graph evaluation")
-	defer func() {
-		span.SetStatus(codes.Ok, "")
+	l.mut.RLock()
+	defer l.mut.RUnlock()
 
-		duration := time.Since(start)
-		level.Info(logger).Log("msg", "finished partial graph evaluation", "duration", duration)
-		l.cm.componentEvaluationTime.Observe(duration.Seconds())
-	}()
+	dependenciesToParentsMap := make(map[dag.Node]dag.Node)
+	for _, parent := range updatedNodes {
+		// Make sure we're in-sync with the current exports of parent.
+		l.cache.CacheExports(parent.ID(), parent.Exports())
+		// We collect all nodes directly incoming to parent.
+		_ = dag.WalkIncomingNodes(l.graph, parent, func(n dag.Node) error {
+			dependenciesToParentsMap[n] = parent
+			return nil
+		})
+	}
 
-	// Make sure we're in-sync with the current exports of c.
-	l.cache.CacheExports(c.ID(), c.Exports())
+	// Submit all dependencies for asynchronous evaluation.
+	// During evaluation, if a node's exports change, Flow will we add it to updated nodes queue (controller.Queue) and
+	// the flow controller will call EvaluateDependencies on it again. This results in a concurrent breadth-first
+	// traversal of the nodes that need to be evaluated.
+	for n, parent := range dependenciesToParentsMap {
+		dependantCtx, span := tracer.Start(spanCtx, "SubmitForEvaluation", trace.WithSpanKind(trace.SpanKindInternal))
+		span.SetAttributes(attribute.String("node_id", n.NodeID()))
+		span.SetAttributes(attribute.String("originator_id", parent.NodeID()))
 
-	_ = dag.WalkReverse(l.graph, []dag.Node{c}, func(n dag.Node) error {
-		if n == c {
-			// Skip over the starting component; the starting component passed to
-			// EvaluateDependencies had its exports changed and none of its input
-			// arguments will need re-evaluation.
-			return nil
+		// Submit the node for asynchronous evaluation.
+		enqueueTime := time.Now()
+		err := l.workerPool.SubmitWithKey(n.NodeID(), l.concurrentEvalFn(n, dependantCtx, tracer, enqueueTime))
+		if err != nil {
+			level.Error(l.log).Log("msg", "failed to submit node for evaluation", "node_id", n.NodeID(), "err", err)
+			span.SetStatus(codes.Error, err.Error())
 		}
+		span.End()
+	}
+
+	// Report queue size metric.
+	l.cm.evaluationQueueSize.Set(float64(l.workerPool.QueueSize()))
```

Comment:
Queue size during my tests remained low. Only sometimes on startup it could get higher (e.g. 30), but would drain quickly.

![image](https://github.com/grafana/agent/assets/17101802/a7db7c01-4312-4344-811b-2cc2c89cca18)